src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Common/PackageCollectionExtensions.cs

```diff
@@ -22,25 +22,25 @@ public static NuGetVersion[] GetPackageVersions(this IEnumerable<PackageIdentity
                 .ToArray();
         }
 
-        public static IEnumerable<IGrouping<string, NuGetVersion>> GroupById(this IEnumerable<PackageIdentity> packages)
+        public static IEnumerable<IGrouping<string, T>> GroupById<T>(this IEnumerable<T> packages) where T : PackageIdentity
         {
             return packages
-                .GroupBy(p => p.Id, p => p.Version, StringComparer.OrdinalIgnoreCase);
+                .GroupBy(p => p.Id, p => p, StringComparer.OrdinalIgnoreCase);
         }
 
-        public static PackageIdentity[] GetLatest(this IEnumerable<PackageIdentity> packages)
+        public static T[] GetLatest<T>(this IEnumerable<T> packages) where T : PackageIdentity
         {
             return packages
                 .GroupById()
-                .Select(g => new PackageIdentity(g.Key, g.MaxOrDefault()))
+                .Select(g => g.OrderByDescending(x => x.Version).FirstOrDefault())
```

Comment:
This could have an undesired effect. It could return an array with one element which is null. If you follow where this code is used, you'll see that there are places where the result of the GetLatest call expects the return to have a list where the element is not null.

The way the code was written, it was always returning an instance of a PackageIdentity object that had its version be null but now this code is going to potentially return a list where the whole T (where T is a PackageIdentity) is null.

Reply:
Yeah, I just wanted to save an allocation to create a new PackageIdentity object and reuse actual object. I believed it's impossible to get empty groups. That way, .Select() statement never runs and an empty array is returned.

Reply:
> It could return an array with one element which is null.

I don't see how this could happen given how lambdas work. The Default will never happen, so the code is redundant, and can be written instead as:

```suggestion
                .Select(g => g.OrderByDescending(x => x.Version).First())
```

Reply:
Fixed. Thanks!